### PR TITLE
Padded SGM

### LINKED ISSUE #316: Implement Padded SGM algorithm
## functionality
Input Graphs G1 and G2 with respective vertex sets {1,...,n_1} and {1,...,n_2} and with respective adjacency matrices A and B, and suppose n_1<n_2, use the SGM function to the padded A and B, then return the restriction of permutation matrix to the vertices of G1.

The output can be compared with Figure 3 in paper seeded graph matching: https://arxiv.org/pdf/1209.0367.pdf.

### PR DESCRIPTION
**Pull request by asaadeldin11**
#### Reference Issues/PRs
Fixes #316, supersedes #472.

#### What does this implement/fix? Explain your changes.
This PR brings padded SGM to graspologic, allowing users to pass in input matrices of different sizes to `GraphMatch`. This is an implementation of the padding scheme describing in [Seeded Graph Matching](https://arxiv.org/pdf/1209.0367.pdf) (section 2.5). Both schemes described (adopted and naive), are implemented here, allowing users the option to choose which they'd prefer to use.

This PR also removes some redundant tests in the match submodule.
#### Any other comments?
Proof of effectiveness can be found in the following [notebook](https://nbviewer.jupyter.org/github/asaadeldin11/nbview/blob/master/padded_sgm.ipynb), where Figure 3 from SGM is replicated.

### CODE REVIEW

**Review by bdpedigo — CHANGES_REQUESTED** [review]
@asaadeldin11 can't tell if this is ready for review since it looks like it was requested automatically, can you lmk when all the build checks are working? Then im guessing we can get this in really quickly

**Comment by asaadeldin11** [comment]
> @asaadeldin11 can't tell if this is ready for review since it looks like it was requested automatically, can you lmk when all the build checks are working? Then im guessing we can get this in really quickly

Yup, not quite ready for review. I'll let you know once I've addressed your previous feedback and have checks working.

File: tests/match/test_match.py [hunk 1]
```diff
@@ -28,20 +32,15 @@ def test_SGM_inputs(self):
             GMP(eps=-1)
         with pytest.raises(TypeError):
             GMP(gmp="hey")
+        with pytest.raises(ValueError):
```
Comment (bdpedigo) [inline tests/match/test_match.py:35]: missed this before - also check for type error for `padding`

**Review by bdpedigo — CHANGES_REQUESTED** [review]
Notebook comments: 
- I still find the use of "core vertices" unclear, for example in "Match ratio of core vertices with naive padding: 1.0
" could you just say "Match ratio of nodes left in G1" or something more eloquent but to that effect?
- looks like something got messed up with the notebook as both plots look the same in netlify render. 
- please note somewhere that the example was adapted from a paper (that's true right?) so we are giving proper attribution 
- super nitpicky but could you change name of files (tutorial and anything else I missed) to `padded_gm` since seeds are not required and that matches the name of the notebook. 

Aside from those minor comments I am happy to merge!

**Comment by bdpedigo** [comment]
@Nyecarr @bryantower @dwaynepryce if anyone can have a look I think this PR is ready for review on MSFT side

**Comment by bdpedigo** [comment]
Github is being weird so i can't re-request a review for myself, but I officially approve.

**Review by bdpedigo — APPROVED** [review]

**Comment by bdpedigo** [comment]
@dwaynepryce any idea why the microsoft checks havent finished running on this? 

also, jovo was hoping to send this code to someone - do you think someone on the MSR side would have time to give this a quick look this week?

File: graspologic/match/gmp.py [hunk 2]
```diff
@@ -130,6 +142,14 @@ def __init__(
         else:
             msg = '"gmp" must be a boolean'
             raise TypeError(msg)
+        if isinstance(padding, str) and padding in {"adopted", "naive"}:
+            self.padding = padding
+        elif isinstance(padding, str):
+            msg = 'Invalid "padding" parameter string'
```
Comment (dwaynepryce) [inline graspologic/match/gmp.py:148]: how did black not complain about using a ' instead of a "

Comment (dwaynepryce) [reply graspologic/match/gmp.py:148]: are you a wizard

**Comment by dwaynepryce** [comment]
This looks fine to me @asaadeldin11, but I did have a question about why `black` didn't complain about you using single ticks on a string.  Regardless if it passes that build step, it's good enough for me, but I *am* confused as to how it did it

**Review by dwaynepryce — APPROVED** [review]

File: graspologic/match/gmp.py (hunk 2 above)
Comment (asaadeldin11) [reply graspologic/match/gmp.py:148]: maybe different ticks? I used " ' " not " ` ".

Comment (bdpedigo) [reply graspologic/match/gmp.py:148]: the exception to blacks rule is strings that have quotes within them like the one here. see https://black.readthedocs.io/en/stable/the_black_code_style.html#strings
though Ali is, in fact, a wizard.

Comment (bdpedigo) [reply graspologic/match/gmp.py:148]: (cause " ... "padding" ..." would break)

**Comment by bdpedigo** [comment]
@dwaynepryce thanks! did you see my notes about the status checks? That is preventing merge

**Comment by dwaynepryce** [comment]
According to the pipeline it never got a request to build.  I found an older one for this PR and told it to run again, so hopefully this works for us?

**Comment by dwaynepryce** [comment]
I have no idea what caused it, but deleting the build pipeline in Azure DevOps and recreating it seems to have made it work again.

Sigh.

**Pull request merged**